libsolidity/interface/StorageLayout.cpp

```diff
@@ -0,0 +1,116 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libsolidity/interface/StorageLayout.h>
+
+#include <libsolidity/ast/TypeProvider.h>
+
+using namespace std;
+using namespace dev;
+using namespace dev::solidity;
+
+Json::Value StorageLayout::generate(ContractDefinition const& _contractDef)
+{
+	solAssert(!m_contract, "");
+	m_contract = &_contractDef;
+	m_types.clear();
+
+	auto typeType = dynamic_cast<TypeType const*>(_contractDef.type());
+	solAssert(typeType, "");
+	auto contractType = dynamic_cast<ContractType const*>(typeType->actualType());
+	solAssert(contractType, "");
+
+	Json::Value variables(Json::arrayValue);
+	for (auto const& var: contractType->stateVariables())
+		variables.append(generate(*get<0>(var), get<1>(var), get<2>(var)));
+
+	Json::Value layout;
+	layout["storage"] = move(variables);
+	layout["types"] = move(m_types);
+	return layout;
+}
+
+Json::Value StorageLayout::generate(VariableDeclaration const& _var, u256 const& _slot, unsigned _offset)
+{
+	Json::Value varEntry;
+	TypePointer varType = _var.type();
+
+	varEntry["label"] = _var.name();
+	varEntry["ast_id"] = to_string(_var.id());
+	varEntry["contract"] = m_contract->name();
+	varEntry["slot"] = _slot.str();
+	varEntry["offset"] = to_string(_offset);
+	varEntry["type"] = typeKeyName(varType);
+
+	generate(varType);
+
+	return varEntry;
+}
+
+void StorageLayout::generate(TypePointer _type)
+{
+	if (m_types.isMember(typeKeyName(_type)))
+		return;
+
+	// Register it now to cut recursive visits.
+	//m_types[typeKeyName(_type)] = {};
```

Comment:
I guess this is commented out because the line below implicitly does it already? So this should be removed?

Reply:
Ah true, will remove